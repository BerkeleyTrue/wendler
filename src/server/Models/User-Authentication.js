import * as R from 'ramda';
import createDebugger from 'debug';
import dedent from 'dedent';
import moment from 'moment';
import { arg, objectType, mutationField } from 'nexus';
import { of, merge } from 'rxjs';
import { normalizeEmail } from 'validator';
import * as OP from 'rxjs/operators';

import renderUserSignInMail from './user-sign-in.js';
import renderUserSignUpMail from './user-sign-up.js';

import { deferPromise, authUtils } from '../utils';

const log = createDebugger('wndlr:server:Models:UserAuthentication');

const createWaitMessage = timeTillAuthReset => dedent`
Please wait at least ${timeTillAuthReset} minute${
  timeTillAuthReset > 1 ? 's' : ''
} for the sign in email to arrive
  before requesting a new one.
`;

export const ttl = moment.duration(15, 'minutes').asMilliseconds();
export const authResetTime = 5;

const UserAuthen = objectType({
  name: 'UserAuthen',
  definition(t) {
    t.string('token');
    t.int('ttl');
    t.int('craeatedAt');
  },
});

const pluckCreatedAt = R.prop('createdAt');
const createResetMoment = () => moment().subtract(authResetTime, 'm');

// (Auth) => Boolean
export const isAuthRecent = R.pipe(
  pluckCreatedAt,
  moment,
  createdOn => createdOn.isAfter(createResetMoment()),
);

// (Auth) => Number
export const getWaitTime = R.pipe(
  pluckCreatedAt,
  moment,
  createdOn => createdOn.diff(createResetMoment()),
  moment.duration,
  dur => dur.minutes(),
);

// () => Observable<Token>
export const createToken = R.pipe(
  authUtils.generateVerificationToken,
  OP.map(token => ({
    token,
    ttl: ttl,
  })),
);

// (Observable<Auth>) => Observable<{ message: string }>
const sendWaitMessageForOldAuth = R.pipe(
  OP.map(getWaitTime),
  OP.map(createWaitMessage),
  OP.map(message => ({ message })),
);

export const createMailSender = (url, sendMail) => ({
  email,
  guid,
  token,
  isSignUp,
}) => {
  const renderText = isSignUp ? renderUserSignUpMail : renderUserSignInMail;

  return sendMail({
    to: email,
    subject: isSignUp ? 'Sign Up' : 'Sign In',
    text: renderText({
      token,
      guid,
      url,
    }),
  });
};

export const sendAuthenEmail = (
  parent,
  { email },
  { get, prisma, sendMail },
) => {
  const url = get('url');
  const sendAuthMail = createMailSender(url, sendMail);
  const normalizedEmail = normalizeEmail(email);
  const findUser = deferPromise(prisma.user.bind(prisma));
  const findAuths = deferPromise(prisma.authenTokens.bind(prisma));
  const createUser = deferPromise(prisma.createUser.bind(prisma));
  const updateUser = deferPromise(prisma.updateUser.bind(prisma));
  const createTokenForUser = deferPromise(
    prisma.createAuthenToken.bind(prisma),
  );

  // email =>
  return R.pipe(
    normalizeEmail,
    // => Observable<User|Void>
    normalizedEmail => findUser({ normalizedEmail }),
    // => Observable<{ user: User|Void, auth: Auth[]|Void }>
    OP.switchMap(user =>
      user ?
        findAuths({ where: { user: { id: user.id } } }).pipe(
          OP.map(auth => ({ user, auth })),
        ) :
        of({ user }),
    ),
    // => [Observable<{ user: User, auth: Auth[]|Void }>, Observable<Void>]
    OP.partition(
      R.pipe(
        R.prop('user'),
        Boolean,
      ),
    ),
    ([
      existingUser,
      noUser,
    ]) => {
      const handleNewUser = R.pipe(
        OP.tap(() => log('new user')),
        // => Token
        OP.switchMap(createToken),
        // => { email, guid, token, isSignUp }
        OP.switchMap(authenToken =>
          createUser({
            email,
            normalizedEmail,
            authenTokens: {
              create: [ authenToken ],
            },
          }).pipe(
            OP.map(({ id, email }) => ({
              email,
              guid: id,
              token: authenToken.token,
              isSignUp: true,
            })),
          ),
        ),
        OP.switchMap(sendAuthMail),
        // OP.tap((emailInfo) => log('emailInfo: ', emailInfo)),
        OP.mapTo({
          message: dedent`
            Welcome! We've sent you a sign in email. Give it a few
            seconds to arrive.
          `,
        }),
      )(noUser);

      // Observable<{ user: User, auth: Auth[]|Void }>
      const handleExistingUser = R.pipe(
        OP.tap(() => log('existing user')),
        // => [
        //  Observable<{ user: User, auth: Auth[] }>,
        //  Observable<{ user: User, auth: Void}>
        // ]
        OP.partition(
          R.pipe(
            R.prop('auth'),
            R.isEmpty,
            R.not,
          ),
        ),
        ([
          // need token split this and check for expired auth before delete
          hasExistingAuth,
          hasNoAuth,
        ]) => {
          const handleNoAuth = R.pipe(
            OP.tap(() => log('no auth')),
            OP.withLatestFrom(createToken()),
            OP.switchMap(([
              { user },
              token,
            ]) =>
              createTokenForUser({
                ...token,
                user: { connect: { id: user.id } },
              }).pipe(
                OP.map(auth => ({
                  email,
                  guid: user.id,
                  isSignUp: false,
                  token: auth.token,
                })),
              ),
            ),
            OP.switchMap(sendAuthMail),
            OP.mapTo({
              message: dedent`
                Sign in email is on it's way!
              `,
            }),
          )(hasNoAuth);

          const handleExistingAuth = R.pipe(
            // => [
            //  Observable<{ user: User, auth: Auth[] }>,
            //  Observable<{ user: User, auth: Auth[] }>
            // ]
            OP.partition(
              R.pipe(
                R.prop('auth'),
                R.head,
                isAuthRecent,
              ),
            ),
            ([
              hasRecentAuth,
              hasOldAuth,
            ]) => {
              const handleOldToken = R.pipe(
                OP.tap(() => log('old token')),
                // create token
                OP.withLatestFrom(createToken()),
                OP.switchMap(([
                  { user, auth },
                  newToken,
                ]) =>
                  updateUser({
                    where: { id: user.id },
                    data: {
                      authenTokens: {
                        delete: auth.map(t => ({ id: t.id })),
                        create: [ newToken ],
                      },
                    },
                  }).pipe(
                    OP.mapTo({
                      guid: user.id,
                      token: newToken.token,
                      email,
                    }),
                  ),
                ),
                OP.switchMap(sendAuthMail),
                OP.mapTo({
                  message: `
                    Sign in
                  `,
                }),
              )(hasOldAuth);

              const handleRecentAuth = R.pipe(
                OP.tap(() => log('has recent auth')),
                OP.pluck('auth'),
                OP.map(R.head),
                sendWaitMessageForOldAuth,
              )(hasRecentAuth);

              return merge(handleOldToken, handleRecentAuth);
            },
          )(hasExistingAuth);

          return merge(handleExistingAuth, handleNoAuth);
        },
      )(existingUser);

      return merge(handleExistingUser, handleNewUser);
    },
    OP.tap(() => {}, err => console.error(err)),
    obv => obv.toPromise(),
  )(email);
};

export const typeDefs = [
  UserAuthen,
  mutationField('sendAuthenEmail', {
    type: 'Info',
    args: {
      email: arg({ type: 'Email' }),
    },
    resolve: sendAuthenEmail,
  }),
];
