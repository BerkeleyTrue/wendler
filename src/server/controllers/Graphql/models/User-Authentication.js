// @flow
import { type Observable } from 'rxjs';
import _ from 'lodash/fp';
import moment from 'moment';

import { authUtils } from '../../../utils';

const ttl15Min = 15 * 60 * 1000;
const authResetTime = 5;

export type AuthenToken = {
  ttl: number,
  token: string,
  createdOn: number,
};

export const isAuthRecent = (createdOn: number) =>
  moment(createdOn).isAfter(moment().subtract(authResetTime, 'm'));

export const getWaitTime = _.flow(
  moment,
  createdOn => createdOn.diff(moment().subtract(authResetTime, 'm')),
  moment.duration,
  _.method('minutes'),
);

export const createToken = (): Observable<AuthenToken> =>
  authUtils.generateVerificationToken().map(token => ({
    token,
    ttl: ttl15Min,
    createdOn: Date.now(),
  }));
