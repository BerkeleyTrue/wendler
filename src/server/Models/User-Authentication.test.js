import moment from 'moment';
import { of } from 'rxjs';
import * as OP from 'rxjs/operators';

import * as UserAthen from './User-Authentication.js';

const createOldAuth = (timeAgo) =>
  UserAthen.createToken().pipe(
    OP.map(authToken => {
      authToken.createdOn = moment()
        .subtract(timeAgo, 'm')
        .valueOf();
      return authToken;
    }),
  );

const createOldAuthFromReset = (addThis) =>
  createOldAuth(UserAthen.authResetTime + addThis);

describe('isAuthRecent', () => {
  test('returns true with recent moment', () =>
    createOldAuthFromReset(-2)
      .pipe(
        OP.map(UserAthen.isAuthRecent),
        OP.tap((isAuthRecent) => expect(isAuthRecent).toBe(true)),
      )
      .toPromise());

  test('returns false with old moment', () =>
    createOldAuthFromReset(2)
      .pipe(
        OP.map(UserAthen.isAuthRecent),
        OP.tap((isAuthRecent) => expect(isAuthRecent).toBe(false)),
      )
      .toPromise());
});

describe('getWaitTime', () => {
  test('returns minutes since moment', () => {
    const momentAgos = UserAthen.authResetTime - 2;
    return createOldAuth(momentAgos)
      .pipe(
        OP.map(UserAthen.getWaitTime),
        OP.tap((waitTime) =>
          expect(waitTime).toBeLessThanOrEqual(
            UserAthen.authResetTime - momentAgos,
          ),
        ),
      )
      .toPromise();
  });
});

describe('sendSignInEmail', () => {
  describe('internals', () => {
    const email = 'foo@dope.business';
    const userPropMatchers = {
      createdOn: expect.any(Number),
      guid: expect.any(String),
      lastUpdatedOn: expect.any(Number),
    };
    const authPropMatchers = {
      createdOn: expect.any(Number),
      token: expect.any(String),
    };
    let queryOne = () => ({});
    let query = () => ({});
    beforeEach(() => {
      queryOne = jest.fn(of);
      query = jest.fn(of);
    });
    describe('queryUserNAuth', () => {
      test('queries using email', () =>
        UserAthen.internals
          .queryUserNAuth(queryOne, email)
          .pipe(
            OP.tap(query => {
              expect(query).toMatchSnapshot();
            }),
          )
          .toPromise());
    });
    describe('createUserAndAuth', () => {
      test('queries using user and auth token', () =>
        UserAthen.internals
          .createUserAndAuth(queryOne, email, of(null))
          .pipe(
            OP.tap(ret => {
              const query = queryOne.mock.calls[0][0];
              expect(query.bindVars.value0).toMatchSnapshot(userPropMatchers);
              expect(query.bindVars.value1).toMatchSnapshot(authPropMatchers);
              expect(query.query).toMatchSnapshot();
              expect(ret).toMatchSnapshot({
                token: expect.any(String),
                guid: expect.any(String),
              });
            }),
          )
          .toPromise());
    });
    describe('createAuthForUser', () => {
      test('queries with old user', () =>
        UserAthen.internals
          .createAuthForUser(
            queryOne,
            of({ user: { _id: 1234, guid: 'asdf' } }),
          )
          .pipe(
            OP.tap(ret => {
              const query = queryOne.mock.calls[0][0];
              expect(ret).toMatchSnapshot({
                token: expect.any(String),
              });
              expect(query.bindVars.value0).toMatchSnapshot(authPropMatchers);
            }),
          )
          .toPromise());
    });
    describe('destroyAndCreateAuthForUser', () => {
      test('queries with old auth', () =>
        UserAthen.internals
          .deleteAndCreateNewAuthForUser(
            query,
            queryOne,
            of({
              user: { _id: 9876, guid: 'asdf' },
              auth: { _id: 1234, _key: 5678 },
            }),
          )
          .pipe(
            OP.tap(ret => {
              const firstQuery = queryOne.mock.calls[0][0];
              const secondQuery = query.mock.calls[0][0];
              expect(ret).toMatchSnapshot({ token: expect.any(String) });
              expect(firstQuery).toMatchSnapshot();
              expect(secondQuery.bindVars).toMatchSnapshot({
                value0: authPropMatchers,
                value1: expect.any(Number),
              });
              expect(secondQuery.query).toMatchSnapshot();
            }),
          )
          .toPromise());
    });
  });
});
