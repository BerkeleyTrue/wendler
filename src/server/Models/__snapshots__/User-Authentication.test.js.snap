// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createMailSender should call sendMail 1`] = `
Object {
  "subject": "Sign Up",
  "text": "Welcome to the WNDRL community!

We have created a new account for you.

Here's your sign in link. It will instantly sign you into WNDLR.com - no password necessary:

foo/passwordless-auth?email=1234&token=asdf

Note: this sign in link will expire after 15 minutes. If you need a new sign in link, go to foo/signin

See you soon!

- The WNDLR Team",
  "to": "Foo",
}
`;

exports[`createMailSender should call sendMail 2`] = `
Object {
  "subject": "Sign In",
  "text": "Here's your sign in link. It will instantly sign you into WNDLR.com - no password necessary:

foo/passwordless-auth?guid=1234&token=asdf

Note: this sign in link will expire after 15 minutes. If you need a new sign in link, go to foo/sign-in

See you soon!

- The WNDLR Team",
  "to": "Foo",
}
`;
